resources/lambdas.md

```diff
@@ -78,23 +78,64 @@ For more on lambda expressions, refer to the
-anything that an object can (be passed as a parameter, returned, etc).
-The main idea behind functional programming is that it's programming without any or with minimal [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). 
-In other words, it's programming where only local variables that are created inside the function can be modified. This means that no global variables, variables passed by reference, or parameterized variables can be
-modified for it to count as functional programming. An example would be:
+In general programming theory, there exists two main types of languages:
+
+- [Imperative programming](https://en.wikipedia.org/wiki/Imperative_programming)
+- [Functional programming](https://en.wikipedia.org/wiki/Functional_programming)
+
+While these two will turn out to seem super different, they are [proven](https://en.wikipedia.org/wiki/Church%E2%80%93Turing_thesis) to be identical.
+
+In a nutshell, functional programming languages treat EVERYTHING as a function, including the basic definitions of `true` and `false`!
+
+From there, every series of actions is not defined as **what something should do**, but rather **what something is** as a recursive set of functions, then asking for a thing.
+While this may seem useless, it is powerful to define incredibly complex processes, and only commit processing power when it is needed (lazy evaluation).
+
+
+```text
+f(x) = y
+
+-- snip --
+
+f(x) -> [what happens in f(x) stays in f(x)...] -> y
+```
+
+From these implementations, two things are realized; namely:
+- Working with sets becomes very useful
+- Functions can't *do anything* besides define more things. (A lot like mathematical functions)
+
+What does that last bit mean? Basically, a function cannot contain [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). In other words, it must follow a
+"black box" model, where all the function/caller can see is the value passed in and the value returned, nothing outside edited, no prints, no outside variables, nada.
+
+Functional programming languages are still widely used, particularly in data science for large processing operations. The YouTube channel Computerphile has a ton of 
+really neat videos on them if you're interested.
+
+## What does this have to do with what I'M working with?
+
+Great question. Well, since around 1993, a lot of imperative language developers started to realize how powerful functional programming can be.
+As you'd expect, Java is one of these functional-capable languages, and you may have even seen some of its functional features already. The most commonly used ones are:
+
+- Anonymous functions (lambdas): functions with no name
+- `map` function: applies a given function to every element in a collection
+- `fold` function: collapses a collection into a single value according to a passed function
```

Comment:
Does this actually exist in Java?
I know in JavaScript it's called reduce. You should probably also include that.